tools/xdp/s2n-quic-xdp/src/if_xdp.rs

```diff
@@ -0,0 +1,265 @@
+        /// application.
+        const USE_NEED_WAKEUP = 1 << 3;
+    }
+);
+
+bitflags!(
+    /// Flags for the umem config
+    ///
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L30
+    #[derive(Copy, Clone, Debug, Default)]
+    #[repr(transparent)]
+    pub struct UmemFlags: u32 {
+        const UNALIGNED_CHUNK_FLAG = 1 << 0;
+    }
+);
+
+/// A structure for representing the address of an AF_XDP socket
+///
+/// https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L32-L38
+#[derive(Clone, Copy, Debug)]
+#[repr(C)]
+pub struct Address {
+    pub family: u16,
+    pub flags: XdpFlags,
+    pub ifindex: u32,
+    pub queue_id: u32,
+    pub shared_umem_fd: u32,
+}
+
+impl Default for Address {
+    fn default() -> Self {
+        Self {
+            family: libc::PF_XDP as _,
+            flags: Default::default(),
+            ifindex: 0,
+            queue_id: 0,
+            shared_umem_fd: 0,
+        }
+    }
+}
+
+impl Address {
+    /// Resolves the `ifindex` for the provided `if_name` and associates it with the address.
+    ///
+    /// If the device does not exist, then an error is returned.
+    #[inline]
+    pub fn set_if_name(&mut self, name: &CStr) -> Result<&mut Self> {
+        unsafe {
+            let ifindex = libc::if_nametoindex(name.as_ptr());
+
+            // https://man7.org/linux/man-pages/man3/if_nametoindex.3.html
+            // > On success, if_nametoindex() returns the index number of the
+            // > network interface; on error, 0 is returned and errno is set to
+            // > indicate the error.
+            if ifindex == 0 {
+                return Err(std::io::Error::last_os_error());
+            }
+
+            self.ifindex = ifindex;
+        }
+        Ok(self)
+    }
+}
+
+bitflags!(
+    /// Flags set on a particular ring state `flags` field
+    ///
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L41
+    #[derive(Default)]
+    #[repr(transparent)]
+    pub struct RingFlags: u32 {
+        const NEED_WAKEUP = 1 << 0;
+    }
+);
+
+/// A structure to communicate the offsets for an individual ring for an AF_XDP socket
+///
+/// Note that the `flags` field was added in kernel 5.4.
+///
+/// See https://github.com/torvalds/linux/blob/9116e5e2b1fff71dce501d971e86a3695acc3dba/include/uapi/linux/if_xdp.h#L28-L32
+#[derive(Clone, Copy, Debug)]
+#[repr(C)]
+pub struct RingOffsetV1 {
+    pub producer: u64,
+    pub consumer: u64,
+    pub desc: u64,
+}
+
+/// A structure to communicate the offsets for an individual ring for an AF_XDP socket
+///
+/// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L43
+#[derive(Clone, Copy, Debug, Default)]
+#[repr(C)]
+pub struct RingOffsetV2 {
```

Comment:
Why do we need a V1? Can we simply claim support for kernel versions after V2 was added?

Reply:
I feel like it's cleaner to support both than to have to probe the version at runtime and have users make sure they're using a supported one. Maybe we can drop support eventually but V1 isn't _that_ old.